crates/runtime/Cargo.toml

```diff
@@ -37,6 +37,7 @@ winapi = { version = "0.3.7", features = ["winbase", "memoryapi", "errhandlingap
 
 [target.'cfg(target_os = "linux")'.dependencies]
 userfaultfd = { version = "0.4.1", optional = true }
+memfd = { version = "0.4.1", optional = true }
```

Comment:
Given where this PR now is and the trajectory of this feature, I think this is good to have as an optional dependency but I think it should be enabled by default and probably just called `memfd` (enabled-by-default probably from the `wasmtime` crate, not here). Especially with this making such a drastric difference for the on-demand allocator it seems like using `memfd` is a no-brainer.

To make this tweak as well as improve the portability here (right now if you enable `memfd` on Windows it'd probably just fail to compile) I think a few small changes are all that's needed. To be clear though this is icing on the cake and it's totally fine to do this as a follow-up, but I think the steps would be:

* Remove the `memfd-allocator`
* Add a `memfd` feature to the `wasmtime` crate that forwards to `wasmtime-runtime/memfd`
* Add `memfd` to the default feature set of the `wasmtime` crate
* Add a `crates/runtime/build.rs` that detects that the target's os is Linux (by looking at `CARGO_CFG_TARGET_OS`) and that the `memfd` feature is active (by looking at `CARGO_FEATURE_MEMFD`), and if both are present then printing `println!("cargo:rustc-cfg=memfd")`
* Change all `cfg(feature = "memfd-allocator")` in the crate to `cfg(memfd)`

With that I think we'll have on-by-default memfd for Linux-only, with the opt-in ability to turn it off. Again though it's fine to defer this to a future PR.

As I type all this out though I also would question maybe we don't even need a feature for this. I suppose one thing that could come up is that if you have a million modules in a process that's a million file descriptors which can blow kernel limits, but other than that I'm not sure why anyone would explicitly want to disable memfd.

Reply:
I've gone ahead and done all of this; the default build now will use memfd, validated by building the `wasmtime` binary and strace'ing an execution. Waiting for CI to see if moving `memfd` out of the target-only deps into the main deps will work on non-Linux, but I am hoping it will (the crate's got a toplevel `#![cfg(target_os ...)]` that should make it a no-op on other platforms...?).